File: app/web/src/api/domain.ts
import apiClient from "./apiClient";

export interface Domain {
	id: number;
	uuid: string;
	name: string;
	icon_url: string;
	description: string;
	owner_uuid: string;
	invite_code: string;
	is_public: boolean;
	created_at: string;
}

export interface DomainDetail extends Domain {
	member_count: number;
	room_count: number;
}

export interface DomainMember {
	id: number;
	domain_uuid: string;
	user_uuid: string;
	nickname: string;
	role_name: string;
	joined_at: string;
	name: string;
	display_name: string;
}

export interface DomainPage {
	domains: Domain[];
	total: number;
}

export interface DomainApiError {
	code?: number;
	msg?: string;
}

export type DomainMutation =
	| "create"
	| "update"
	| "delete"
	| "join"
	| "leave"
	| "kick";

export async function createDomain(data: {
	name: string;
	description?: string;
	is_public?: boolean;
}): Promise<Domain> {
	const result = await apiClient.post<Domain>({
		url: "/api/v1/domain/create",
		data,
	});
	if (!result) throw new Error("domain data is missing");
	return result;
}

export async function getDomain(uuid: string): Promise<Domain> {
	const result = await apiClient.post<Domain>({
		url: "/api/v1/domain/get",
		data: { domain_uuid: uuid },
	});
	if (!result) throw new Error("domain data is missing");
	return result;
}

export async function listDomains(
	page = 1,
	pageSize = 20,
): Promise<DomainPage> {
	const result = await apiClient.post<DomainPage>({
		url: "/api/v1/domain/list",
		data: { page, page_size: pageSize },
	});
	if (!result) throw new Error("domain data is missing");
	return result;
}

export async function listPublicDomains(
	page = 1,
	pageSize = 20,
	keyword?: string,
): Promise<DomainPage> {
	const data: { page: number; page_size: number; keyword?: string } = {
		page,
		page_size: pageSize,
	};
	if (keyword?.trim()) {
		data.keyword = keyword.trim();
	}
	const result = await apiClient.post<DomainPage>({
		url: "/api/v1/domain/list-public",
		data,
	});
	if (!result) throw new Error("domain data is missing");
	return result;
}

export async function getMyDomainsDetailed(): Promise<DomainDetail[]> {
	const result = await apiClient.post<DomainDetail[]>({
		url: "/api/v1/domain/my-domains",
		data: {},
	});
	if (!result) throw new Error("domain data is missing");
	return result;
}

export async function myDomains(): Promise<string[]> {
	const details = await getMyDomainsDetailed();
	return details.map((detail) => detail.uuid);
}

export async function previewDomainInvite(inviteCode: string): Promise<Domain> {
	const result = await apiClient.post<Domain>({
		url: "/api/v1/domain/preview",
		data: { invite_code: inviteCode },
	});
	if (!result) throw new Error("domain data is missing");
	return result;
}

export async function joinDomain(inviteCode: string): Promise<Domain> {
	const result = await apiClient.post<Domain>({
		url: "/api/v1/domain/join",
		data: { invite_code: inviteCode },
	});
	if (!result) throw new Error("domain data is missing");
	return result;
}

export async function leaveDomain(uuid: string): Promise<void> {
	await apiClient.post({
		url: "/api/v1/domain/leave",
		data: { domain_uuid: uuid },
	});
}

export async function updateDomain(data: {
	domain_uuid: string;
	name?: string;
	description?: string;
	icon_url?: string;
	is_public?: boolean;
}): Promise<Domain> {
	const result = await apiClient.post<Domain>({
		url: "/api/v1/domain/update",
		data,
	});
	if (!result) throw new Error("domain data is missing");
	return result;
}

export async function deleteDomain(uuid: string): Promise<void> {
	await apiClient.post({
		url: "/api/v1/domain/delete",
		data: { domain_uuid: uuid },
	});
}

export async function kickDomainMember(
	domainUUID: string,
	userUUID: string,
): Promise<void> {
	await apiClient.post({
		url: "/api/v1/domain/kick",
		data: { domain_uuid: domainUUID, user_uuid: userUUID },
	});
}

export async function domainMembers(
	domainUUID: string,
): Promise<DomainMember[]> {
	const data = await apiClient.post<{ members: DomainMember[] }>({
		url: "/api/v1/domain/members",
		data: { domain_uuid: domainUUID },
	});
	return data?.members ?? [];
}

File: app/server/internal/model/domain.go
package model

import (
	"time"

	"github.com/nrednav/cuid2"
	"gorm.io/gorm"
)

// Domain 代表一个语音域（原 Domain/Server 语义）。
type Domain struct {
	ID          uint      `gorm:"primaryKey" json:"id"`
	UUID        string    `gorm:"size:32;uniqueIndex" json:"uuid"`
	Name        string    `gorm:"size:100;not null" json:"name"`
	IconURL     string    `gorm:"size:512" json:"icon_url"`
	Description string    `gorm:"size:500" json:"description"`
	OwnerUUID   string    `gorm:"type:uuid;index;not null" json:"owner_uuid"`
	InviteCode  string    `gorm:"size:32;uniqueIndex" json:"invite_code"`
	IsPublic    bool      `gorm:"default:false" json:"is_public"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

func (d *Domain) TableName() string {
	return "domains"
}

func (d *Domain) BeforeCreate(_ *gorm.DB) error {
	if d.UUID == "" {
		d.UUID = cuid2.Generate()
	}
	if d.InviteCode == "" {
		d.InviteCode = generateInviteCode()
	}
	return nil
}

// DomainMember 用户-Domain 多对多关系。
type DomainMember struct {
	ID         uint      `gorm:"primaryKey" json:"id"`
	DomainUUID string    `gorm:"size:32;index:idx_domain_member,priority:1;not null" json:"domain_uuid"`
	UserUUID   string    `gorm:"type:uuid;index:idx_domain_member,priority:2;not null" json:"user_uuid"`
	Nickname   string    `gorm:"size:64" json:"nickname"`
	RoleName   string    `gorm:"size:32;default:member" json:"role_name"`
	JoinedAt   time.Time `json:"joined_at"`
}

func (DomainMember) TableName() string {
	return "domain_members"
}

// DomainMemberInfo 成员列表视图，附带用户全局名称。
type DomainMemberInfo struct {
	ID          uint      `json:"id"`
	DomainUUID  string    `json:"domain_uuid"`
	UserUUID    string    `json:"user_uuid"`
	Nickname    string    `json:"nickname"`
	RoleName    string    `json:"role_name"`
	JoinedAt    time.Time `json:"joined_at"`
	Name        string    `json:"name"`
	DisplayName string    `json:"display_name"`
}

// DomainDetail 我的语音域批量详情视图，附带成员数与房间数。
type DomainDetail struct {
	ID          uint      `json:"id"`
	UUID        string    `json:"uuid"`
	Name        string    `json:"name"`
	IconURL     string    `json:"icon_url"`
	Description string    `json:"description"`
	OwnerUUID   string    `json:"owner_uuid"`
	InviteCode  string    `json:"invite_code"`
	IsPublic    bool      `json:"is_public"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	MemberCount int64     `json:"member_count"`
	RoomCount   int64     `json:"room_count"`
}

// generateInviteCode 生成 8 字符随机邀请码。
func generateInviteCode() string {
	const charset = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
	u := cuid2.Generate()
	var b [8]byte
	for i := range b {
		c := u[i]
		var v byte
		switch {
		case c >= '0' && c <= '9':
			v = c - '0'
		case c >= 'a' && c <= 'z':
			v = c - 'a' + 10
		default:
			v = 0
		}
		b[i] = charset[int(v)%len(charset)]
	}
	return string(b[:])
}
